protobuf/key_attributes.proto

```diff
@@ -18,172 +18,78 @@ syntax = "proto3";
-  MD4 = 2;
-  MD5 = 3;
-  RIPEMD160 = 4;
-  SHA_1 = 5;
-  SHA_224 = 6;
-  SHA_256 = 7;
-  SHA_384 = 8;
-  SHA_512 = 9;
-  SHA_512_224 = 10;
-  SHA_512_256 = 11;
-  SHA3_224 = 12;
-  SHA3_256 = 13;
-  SHA3_384 = 14;
-  SHA3_512 = 15;
-}
-
-enum CipherAlgorithm {
-  ARC4 = 0;
-  CTR = 1;
-  CFB = 2;
-  OFB = 3;
-  XTS = 4;
-  CBC_No_Padding = 5;
-  CBC_PKCS7 = 6;
-}
-
-enum AsymmetricEncryptionAlgorithm {
-  RSA_PKCS1V15_Crypt = 0;
-  RSA_OAEP = 1;
-}
-
-message AsymmetricEncryption {
-  AsymmetricEncryptionAlgorithm asymmetric_encryption_algorithm = 1;
-  HashAlgorithm hash_algorithm = 2;
-}
-
-enum MacAlgorithm {
-  HMAC = 0;
-  CBC_MAC = 1;
-  CMAC = 2;
-  GMAC = 3;
-}
-
-message Mac {
-  MacAlgorithm mac_algorithm = 1;
-  HashAlgorithm hash_algorithm = 2;
-  uint32 tag_length = 3;
-}
-
-enum AeadAlgorithm {
-  CCM = 0;
-  GCM = 1;
-}
-
-message Aead {
-  AeadAlgorithm aead_algorithm = 1;
-  uint32 tag_length = 2;
-}
-
-enum SignAlgorithm {
-  RSA_PKCS1V15_Sign = 0;
-  RSA_PSS = 1;
-  DSA = 2;
-  Deterministic_DSA = 3;
-  ECDSA = 4;
-  Deterministic_ECDSA = 5;
-}
-
-message Sign {
-  SignAlgorithm sign_algorithm = 1;
-  HashAlgorithm hash_algorithm = 2;
-}
-
-enum KeyDerivationFunction {
-  HKDF = 0;
-  TLS12_PRF = 1;
-  TLS12_PSK_To_MS = 2;
-  Select_Raw = 3;
-}
-
-message KeyDerivation {
-  KeyDerivationFunction key_derivation_function = 1;
-  HashAlgorithm hash_algorithm = 2;
-}
-
-enum KeyAgreementAlgorithm {
-  FFDH = 0;
-  ECDH = 1;
+import "algorithm.proto";
+
+message KeyAttributes {
+  KeyType key_type = 1;
+  uint32 key_bits = 2;
+  KeyPolicy key_policy = 3;
+}
+
+message KeyType {
+  message RawData {}
+  message Hmac {}
+  message Derive {}
+  message Aes {}
+  message Des {}
+  message Camellia {}
+  message Arc4 {}
+  message Chacha20 {}
+  message RsaPublicKey {}
+  message RsaKeyPair {}
+  message EccKeyPair { EccFamily curve_family = 1; }
+  message EccPublicKey { EccFamily curve_family = 1; }
+  message DhKeyPair { DhFamily group_family = 1; }
+  message DhPublicKey { DhFamily group_family = 1; }
+  enum EccFamily {
+    ECC_FAMILY_NONE = 0; // This default variant should not be used.
+    SECP_K1 = 1;
+    SECP_R1 = 2;
+    SECP_R2 = 3 [ deprecated = true ];
+    SECT_K1 = 4; // DEPRECATED for sect163k1 curve
+    SECT_R1 = 5; // DEPRECATED for sect163r1 curve
+    SECT_R2 = 6 [ deprecated = true ];
+    BRAINPOOL_P_R1 = 7; // DEPRECATED for brainpoolP160r1 curve
+    FRP = 8;
+    MONTGOMERY = 9;
+  }
+  enum DhFamily { RFC7919 = 0; }
+
+  oneof variant {
+    // Symmetric keys
+    RawData raw_data = 1;
+    Hmac hmac = 2;
+    Derive derive = 3;
+    Aes aes = 4;
+    Des des = 5 [ deprecated = true ];
+    Camellia camellia = 6;
+    Arc4 arc4 = 7 [ deprecated = true ];
+    Chacha20 chacha20 = 8;
+    // RSA keys
+    RsaPublicKey rsa_public_key = 9;
+    RsaKeyPair rsa_key_pair = 10;
+    // Elliptic Curve keys
+    EccKeyPair ecc_key_pair = 11;
+    EccPublicKey ecc_public_key = 12;
+    // Diffie Hellman keys
```

Comment:
Just to avoid the confusion I went through, could you please add "Finite field" here, so `Finite field Diffie Hellman keys`

Reply:
Sure!